// @evilmartians/agent-prism-data/dist/open-telemetry/utils/open-telemetry-category-mappers.js
import { STANDARD_OPENTELEMETRY_ATTRIBUTES, STANDARD_OPENTELEMETRY_PATTERNS, } from "@evilmartians/agent-prism-types";
import { getOpenTelemetryAttributeValue } from "./get-open-telemetry-attribute-value.js";
export const openTelemetryCategoryMappers = {
    isHttpCall: (span) => {
        return (getOpenTelemetryAttributeValue(span, STANDARD_OPENTELEMETRY_ATTRIBUTES.HTTP_METHOD) !== undefined);
    },
    isDatabaseCall: (span) => {
        return (getOpenTelemetryAttributeValue(span, STANDARD_OPENTELEMETRY_ATTRIBUTES.DB_SYSTEM) !== undefined);
    },
    isFunctionCall: (span) => {
        const name = span.name.toLowerCase();
        return (STANDARD_OPENTELEMETRY_PATTERNS.FUNCTION_KEYWORDS.some((keyword) => name.includes(keyword)) ||
            getOpenTelemetryAttributeValue(span, STANDARD_OPENTELEMETRY_ATTRIBUTES.FUNCTION_NAME) !== undefined);
    },
    isLLMCall: (span) => {
        const name = span.name.toLowerCase();
        return STANDARD_OPENTELEMETRY_PATTERNS.LLM_KEYWORDS.some((keyword) => name.includes(keyword));
    },
    isChainOperation: (span) => {
        const name = span.name.toLowerCase();
        return STANDARD_OPENTELEMETRY_PATTERNS.CHAIN_KEYWORDS.some((keyword) => name.includes(keyword));
    },
    isAgentOperation: (span) => {
        const name = span.name.toLowerCase();
        return STANDARD_OPENTELEMETRY_PATTERNS.AGENT_KEYWORDS.some((keyword) => name.includes(keyword));
    },
    isRetrievalOperation: (span) => {
        const name = span.name.toLowerCase();
        return STANDARD_OPENTELEMETRY_PATTERNS.RETRIEVAL_KEYWORDS.some((keyword) => name.includes(keyword));
    },
};
